Define NDEBUG in release mode

gassmoeller opened the pull request:
Let me start by saying I am not fully aware of the impact of this. This is more of a question: By default cmake would define NDEBUG in anything but debug mode, to disable asserts in the standard library. Deal.II overwrites this default in the `DEAL_II_INITIALIZE_CACHED_VARIABLES` macro, when it resets CMAKE_CXX_FLAGS and it never defines it itself (it instead uses DEBUG to activate asserts in debug mode). I am not sure if this was made on purpose, but inside ASPECT we have some downstream projects (specifically https://github.com/GeodynamicWorldBuilder/WorldBuilder) that rely on NDEBUG to decide whether to activate asserts or not. Does it make sense to keep cmake's default behavior of defining NDEBUG in release mode? This may even give some performance benefits, but I do not know if it was removed no purpose.

drwells commented:
I think external projects should set themselves (i.e., we should not add it to our list of release flags) - for example, if you do `-DCMAKE_BUILD_TYPE=Release`, `TARGET_LINK_LIBRARIES(myproject PUBLIC ${DEAL_LIBRARIES_RELEASE})` etc. then CMake will define `NDEBUG` during compilation - this is what I would expect to happen. If you set your own `CMAKE_CXX_FLAGS` then this behavior is overridden and CMake sets whatever flags you specify.

Does WorldBuilder just set their flags equal to deal.II's flags?

tamiko commented:
@gassmoeller To answer your first question: We override CMake's default choices for debug and release builds because our autoconf/make based build system back in the day had a different set of compiler flags and compilation logic.

Regarding `NDEBUG` - we could definitely set it for our release builds.

I think the major reason why we don't already do that is the simply reason that we don't use `assert()` internally so it hardly matters for us: Debug asserts for the standard library itself are controlled via `-D_GLIBCXX_DEBUG` (libstdc++) or `-D_LIBCPP_DEBUG` (clang's libc++) - and for calling into trilinos/petsc/etc. `NDEBUG` doesn't play a role either.

gassmoeller commented:
> Does WorldBuilder just set their flags equal to deal.II's flags?

We compile WorldBuilder with the ASPECT build system and that one just takes the deal.II flags. Yes, we could change ASPECT's build system to define its own flags but so far is has worked well for us. For now we fixed the whole issue by adding NDEBUG in ASPECT's build system.

> I think the major reason why we don't already do that is the simply reason that we don't use assert() internally so it hardly matters for us

Yeah I thought so. I just wanted to make sure it was not disabled on purpose for some reason.

> Regarding NDEBUG - we could definitely set it for our release builds.

As long as there is no reason not to do it then I think it may be useful to set it just to avoid deviating from cmake standard values. Would the way I implemented it here be correct?

drwells reviewed (approved):
> As long as there is no reason not to do it then I think it may be useful to set it just to avoid deviating from cmake standard values.

That's a good point - I am now convinced that we should do this. The way this patch sets `NDEBUG` is equivalent to how we set `DEBUG` in debug mode so I believe this is the correct approach.

tamiko commented on cmake/setup_compiler_flags_gnu.cmake:
```diff
@@ -187,6 +187,12 @@ IF (CMAKE_BUILD_TYPE MATCHES "Release")
   ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS_RELEASE "-fstrict-aliasing")
 
+  #
+  # Disable assertions in the standard library in release mode
```
Would you mind changing this to `Disable assert() in deal.II and user projects in release mode`?
Because, technically speaking, `-DNDEBUG` is not controlling assertions in the standard library itself.

tamiko commented on cmake/setup_compiler_flags_intel.cmake:
```diff
@@ -183,6 +183,12 @@ IF (CMAKE_BUILD_TYPE MATCHES "Release")
+  # Disable assertions in the standard library in release mode
+  #
+  LIST(APPEND DEAL_II_DEFINITIONS_RELEASE "NDEBUG")
+  LIST(APPEND DEAL_II_USER_DEFINITIONS_RELEASE "NDEBUG")
```
as above

tamiko reviewed (approved):
Good to go! (some minor rewording request).

gassmoeller commented:
I modified the comments. Let's see what the testers say.

tamiko commented:
/rebuild
